crates/stress-test/src/main.rs

```diff
@@ -33,135 +45,181 @@ struct Cli {
 
-        tracker.spawn(async move {
-            // take a permit as this section might have to run serially
-            let mut permit = Some(semaphore.acquire().await?);
+            // wait for all tasks to be set up
+            barrier.wait().await;
 
-            task.create(container_output).await?;
-            permit.drop_if(serial_steps == Step::Create);
+            // Wait for a concurrentcy slot
+            let permit = semaphore.acquire_owned().await?;
+            let _ = start.set(Instant::now());
 
+            task.create(container_output).await?;
             task.start().await?;
-            permit.drop_if(serial_steps == Step::Start);
 
-            task.wait().await?;
-            permit.drop_if(serial_steps == Step::Wait);
+            // release the concurrency slot
+            drop(permit);
 
+            task.wait().await?;
             task.delete().await?;
-            permit.drop_if(serial_steps == Step::Delete);
 
             Ok(())
         });
     }
 
-    println!("Waiting for tasks to finish.");
-    println!("Press Ctrl-C to terminate.");
+    let setup_done = barrier.wait().fuse();
+    let mut setup_done = pin!(setup_done);
 
+    eprintln!("> Setting up tasks.");
+    eprintln!("  Press Ctrl-C to terminate.\x1b[A");
+
+    let mut incomplete = count;
     let mut success = 0;
     let mut failed = 0;
-    let ping = Arc::new(Notify::new());
-    let _ = async {
-        let ping = ping.clone();
-        let count = count as usize;
-        while let Some(res) = tracker.join_next().await {
-            ping.notify_waiters();
-            match res {
-                Ok(Ok(())) => {
-                    success += 1;
-                    println!(" > {} .. [OK]", count - tracker.len());
-                }
-                Ok(Err(err)) => {
-                    failed += 1;
-                    println!(" > {} .. {err}", count - tracker.len());
-                }
-                Err(err) => {
-                    println!(" > {} .. {err}", count - tracker.len());
+
+    loop {
+        tokio::select! {
+            _ = &mut setup_done => {
+                eprint!("\x1b[2K");
+                eprintln!("> Waiting for tasks to finish.");
+                eprintln!("  Press Ctrl-C to terminate.\x1b[A");
+            }
+            _ = watchdog(timeout), if setup_done.is_terminated() => {
+                eprintln!("\x1b[2K");
+                eprintln!("\x1b[31mTimeout\x1b[0m");
+                break;
+            }
+            _ = ctrl_c() => {
+                eprintln!("\x1b[2K");
+                eprintln!("\x1b[31mCancelled\x1b[0m");
+                break;
+            }
+            res = tracker.next() => {
+                let Some(res): Option<Result<()>> = res else {
+                    break;
+                };
+                match res {
+                    Ok(()) => {
+                        incomplete -= 1;
+                        success += 1;
+                        if verbose {
```

Comment:
I'd love this to print by default because this will give me a sense of progress :)

Reply:
I think we can show progress without cluttering the command line :-)